File: app/src/main/java/com/example/basiclist/GameListAdapter.java
package com.example.basiclist;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

public class GameListAdapter extends BaseAdapter {

    Context context;
    LayoutInflater inflater;
    List<Game> gameList;
    ArrayList<Game> gameArrayList;

    public GameListAdapter(Context mContext, List<Game> gameList) {
        context = mContext;
        this.gameList = gameList;
        inflater = LayoutInflater.from(context);
        gameArrayList = new ArrayList<>(gameList);
    }

    public class ViewHolder {
        TextView mTitle, mDescription;
    }

    @Override
    public int getCount() {
        return gameList.size();
    }

    @Override
    public Object getItem(int position) {
        return gameList.get(position);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        ViewHolder viewHolder;
        if (convertView == null) {
            viewHolder = new ViewHolder();
            convertView = inflater.inflate(R.layout.custom_list, null);

            viewHolder.mTitle = convertView.findViewById(R.id.title);
            viewHolder.mDescription = convertView.findViewById(R.id.description);

            convertView.setTag(viewHolder);
        } else {
            viewHolder = (ViewHolder) convertView.getTag();
        }

        viewHolder.mTitle.setText(gameList.get(position).getTitle());
        viewHolder.mDescription.setText(gameList.get(position).getDescription());

        return convertView;
    }

    public void filter(String searchText) {
        searchText = searchText.toLowerCase();
        gameList.clear();
        if (searchText.length() == 0) {
            gameList.addAll(gameArrayList);
        } else {
            for (Game game : gameArrayList) {
                if (game.getTitle().toLowerCase().contains(searchText)) {
                    gameList.add(game);
                }
            }
        }
        notifyDataSetChanged();
    }
}

File: app/src/main/java/com/example/basiclist/GameListAdapter.kt
package com.example.basiclist

import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.Filter
import android.widget.Filterable
import androidx.recyclerview.widget.RecyclerView
import com.example.basiclist.databinding.CustomListBinding

class GameListAdapter(private val data: List<Game>) :
    RecyclerView.Adapter<GameListAdapter.ViewHolder>(), Filterable {

    var gameFilterList = data;

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder.from(parent)
    }


    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val game = gameFilterList[position]
        holder.bind(game)
    }


    override fun getFilter(): Filter {
        return object : Filter() {
            override fun performFiltering(constraint: CharSequence?): FilterResults {
                val resultList = ArrayList<Game>()
                if (constraint.isNullOrEmpty()) {
                    resultList.addAll(data)
                } else {
                    for (game in data) {
                        if ((game.getTitle().toLowerCase() + game.getDescription().toLowerCase())
                                .contains(constraint.toString().toLowerCase())
                        ) {
                            resultList.add(game)
                        }
                    }
                }
                val filterResults = FilterResults()
                filterResults.values = resultList
                return filterResults
            }

            override fun publishResults(constraint: CharSequence?, results: FilterResults?) {
                gameFilterList = results?.values as List<Game>
                notifyDataSetChanged()
            }

        }
    }


    override fun getItemCount() = gameFilterList.size

    class ViewHolder(private val binding: CustomListBinding) :
        RecyclerView.ViewHolder(binding.root) {

        fun bind(game: Game) {
            binding.title.text = game.title
            binding.description.text = game.description
        }

        companion object {
            fun from(parent: ViewGroup): ViewHolder {
                val layoutInflater = LayoutInflater.from(parent.context)
                val binding = CustomListBinding.inflate(layoutInflater, parent, false)
                return ViewHolder(binding)
            }
        }
    }
}